src/qofono.cpp

```diff
@@ -403,8 +1928,24 @@ QOfono::QOfono()
 
 QString QOfono::mobileCountryCodeToAlpha2CountryCode(int mcc)
 {
-    if (mcc < 0 || mcc >= COUNTRY_CODES_LENGTH) {
-        qWarning() << Q_FUNC_INFO << "given unknown Mobile Country Code:" << mcc;
+    int low = 0;
+    int high = (sizeof(qofonoMccList)/sizeof(qofonoMccList[0]))-1;
+
+    while (low <= high) {
+        int mid = (low + high)/2;
+        int val = qofonoMccList[mid].mcc;
+        if (val < mcc) {
+            low = mid + 1;
+        } else if (val > mcc) {
+            high = mid - 1;
+        } else {
+            /* MCC is found. Back up a litte to the first entry */
+            while (mid > 0 && qofonoMccList[mid-1].mcc == val) mid--;
+            return QString(qofonoMccList[mid].cc);
```

Comment:
Probably does not make any difference in one direction or another, but you /can/ avoid the linear search backwards by adjusting the treatment of high bound and continuing the binary search to conclusion. Something like this (smoketested in c app):

```
const char *mobileCountryCodeToAlpha2CountryCode(int mcc)
{
    size_t N  = sizeof qofonoMccList / sizeof *qofonoMccList;
    size_t lo = 0;
    size_t hi = N;
    while( lo < hi ) {
	size_t i = (lo + hi) / 2;
	if( qofonoMccList[i].mcc >= mcc )
	    hi = i;
	else
	    lo = i + 1;
    }
    /* If there are matching entries in the lookup table, the
     * 'hi' is left pointing to the 1st one of them */
    if(hi < N && qofonoMccList[hi].mcc == mcc)
	return qofonoMccList[hi].cc;
    return "";
}
```